[SPARK-29922][SQL] SHOW FUNCTIONS should do multi-catalog resolution

planga82 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Add ShowFunctionsStatement and make SHOW FUNCTIONS go through the same catalog/table resolution framework of v2 commands.

We don’t have this methods in the catalog to implement an V2 command
* catalog.listFunctions

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
It's important to make all the commands have the same table resolution behavior, to avoid confusing
`SHOW FUNCTIONS LIKE namespace.function`

### Does this PR introduce any user-facing change?
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->
Yes. When running SHOW FUNCTIONS LIKE namespace.function Spark fails the command if the current catalog is set to a v2 catalog.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Unit tests.

planga82 commented:
cc: @cloud-fan @rdblue @viirya @imback82

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #114445 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114445/testReport)** for PR 26667 at commit [`7007274`](https://github.com/apache/spark/commit/7007274756d57287d3c11202cf43d7aa9137582b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19281/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:
```diff
@@ -475,3 +475,17 @@ case class ShowCurrentNamespaceStatement() extends ParsedStatement
+ *  SHOW FUNCTIONS statement, as parsed from SQL
+ */
+case class ShowFunctionsStatement(
+    scope: Option[String],
```
shall we simply have 2 boolean flags `userScope` and `systemScope`?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -455,6 +455,25 @@ class ResolveSessionCatalog(
+      import ShowFunctionsStatement._
+      val userScope = scope.map(s => s == ALL || s == USER).getOrElse(true)
+      val systemScope = scope.map(s => s == ALL || s == SYSTEM).getOrElse(true)
+      val (db, function) = functionName match {
```
the function name may contain catalog name, we should resolve it first.  We should match `ShowFunctionsStatement(..., SessionCatalog(_, functionName))`

planga82 replied:
Thanks, I was a little confused with other pull request

imback82 replied:
Looks like this is not addressed as suggested by @cloud-fan?

planga82 replied:
I am in doubt with this in the case that the catalog is not a session catalog. We have to resolve catagog, but if the catalog is not a session catalog I think we have to show a good error message. When we do the same in other commands we do something like this to generate an exception
` if (!isSessionCatalog(catalog)) {
      throw new AnalysisException(s"$sql is only supported with v1 tables.")
    }`
If we do something like this
`case ShowFunctionsStatement(..., SessionCatalog(_, functionName))`
It becomes an unresolved exception because any case match,  and that error is not so clear.

This is the reason why I do in that way but it's posible I'm missing something.

cloud-fan replied:
Currently the table loop up here doesn't respect multi-catalog. e.g. what happens for `SHOW FUNCTIONS myCata.ns1`?

You have a point that we should fail nicely for v2 catalogs. How about
```
case ShowFunctionsStatement(userScope, systemScope, pattern,
    CatalogAndIdentifierParts(catalog, functionName)) =>
  if (isSessionCatalog(catalog)) ... else fail...
```

planga82 replied:
It can't be exactly in that way because functionName is an option. What do you think if we do it as the last changes I have pushed?

SparkQA commented:
**[Test build #114445 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114445/testReport)** for PR 26667 at commit [`7007274`](https://github.com/apache/spark/commit/7007274756d57287d3c11202cf43d7aa9137582b).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114445/
Test FAILed.

SparkQA commented:
**[Test build #114479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114479/testReport)** for PR 26667 at commit [`41e2c3c`](https://github.com/apache/spark/commit/41e2c3c16e396f7f24f959d278f067a63f3a0923).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19313/
Test PASSed.

SparkQA commented:
**[Test build #114479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114479/testReport)** for PR 26667 at commit [`41e2c3c`](https://github.com/apache/spark/commit/41e2c3c16e396f7f24f959d278f067a63f3a0923).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114479/
Test FAILed.

SparkQA commented:
**[Test build #114488 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114488/testReport)** for PR 26667 at commit [`6c12465`](https://github.com/apache/spark/commit/6c124651202969c6de63d658540ea8fff960a1ba).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19321/
Test PASSed.

imback82 commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/statements.scala:
```diff
@@ -475,3 +475,18 @@ case class ShowCurrentNamespaceStatement() extends ParsedStatement
+    functionName: Option[Seq[String]]) extends ParsedStatement
+
+object ShowFunctionsStatement{
+  val ALL = "all"
```
Since these constants are used only in `visitShowFunctions`, does it make sense to define these here? I think the original implementation is also fine?

cloud-fan replied:
We can inline them in the parser rule `visitShowFunctions`

SparkQA commented:
**[Test build #114488 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114488/testReport)** for PR 26667 at commit [`6c12465`](https://github.com/apache/spark/commit/6c124651202969c6de63d658540ea8fff960a1ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114488/
Test FAILed.

planga82 committed:
[SPARK-29922][SQL] Parser Modifications


[SPARK-29922][SQL] Analyzer modifications


[SPARK-29922] Resolve catalog


[SPARK-29922] Fix problems

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19381/
Test PASSed.

SparkQA commented:
**[Test build #114552 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114552/testReport)** for PR 26667 at commit [`7c18a22`](https://github.com/apache/spark/commit/7c18a2293c8f91cdc9a403aff41116b8f691e8d9).

SparkQA commented:
**[Test build #114552 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114552/testReport)** for PR 26667 at commit [`7c18a22`](https://github.com/apache/spark/commit/7c18a2293c8f91cdc9a403aff41116b8f691e8d9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class ShowFunctionsStatement(`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114552/
Test FAILed.

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3273,4 +3273,23 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+  override def visitShowFunctions(ctx: ShowFunctionsContext): LogicalPlan = withOrigin(ctx) {
+    val ALL = "all"
+    val SYSTEM = "system"
+    val USER = "user"
```
defining these variables isn't better. they're not reused too.

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -455,6 +455,23 @@ class ResolveSessionCatalog(
+                throw new AnalysisException(s"Unsupported function name '${functionName.quoted}'")
+            }
+          } else {
+            throw new AnalysisException (s"SHOW FUNCTIONS is only supported in v1 catalog")
```
we don't need s"".

viirya commented on sql/core/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveSessionCatalog.scala:
```diff
@@ -455,6 +455,23 @@ class ResolveSessionCatalog(
+                throw new AnalysisException(s"Unsupported function name '${functionName.quoted}'")
+            }
+          } else {
+            throw new AnalysisException (s"SHOW FUNCTIONS is only supported in v1 catalog")
```
this part is duplicate to parseV1Table.

planga82 replied:
It is not exactly the same message 
$sql is only supported with v1 **tables**.
vs
SHOW FUNCTIONS is only supported in v1 **catalog**
Here we don't have tables so this is the reason the message is different.

viirya commented on sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2SQLSuite.scala:
```diff
@@ -1786,6 +1786,15 @@ class DataSourceV2SQLSuite
+    val e = intercept[AnalysisException] {
+      sql(s"SHOW FUNCTIONS LIKE $function")
+    }
+    assert(e.message.contains(s"SHOW FUNCTIONS is only supported in v1 catalog"))
```
we don't need s"".

cloud-fan commented:
LGTM except the comments from @viirya

planga82 commented:
I'm a few days out, I will review the comments. I was reviewing the tests because I'm not sure why compatibility tests are failing.
Thanks

planga82 committed:
[SPARK-29922] Pull request comments

SparkQA commented:
**[Test build #114877 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114877/testReport)** for PR 26667 at commit [`a38bc74`](https://github.com/apache/spark/commit/a38bc7451ba7149d1c931e0bd03cb4478c5468d2).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19700/
Test PASSed.

SparkQA commented:
**[Test build #114877 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114877/testReport)** for PR 26667 at commit [`a38bc74`](https://github.com/apache/spark/commit/a38bc7451ba7149d1c931e0bd03cb4478c5468d2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114877/
Test FAILed.

planga82 committed:
[SPARK-29922] Test commit to trace better the problems

planga82 committed:
Revert "[SPARK-29922] Test commit to trace better the problems"

This reverts commit 17fb0468aecff69d584b036f23f654192a3fbea2.

SparkQA commented:
**[Test build #114946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114946/testReport)** for PR 26667 at commit [`17fb046`](https://github.com/apache/spark/commit/17fb0468aecff69d584b036f23f654192a3fbea2).

planga82 commented:
Do not review or merge until I find the test problems. Running in local the test finishes without errors. I'm testing a versión in Jenkins to get more information about the error

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19767/
Test PASSed.

SparkQA commented:
**[Test build #114946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114946/testReport)** for PR 26667 at commit [`17fb046`](https://github.com/apache/spark/commit/17fb0468aecff69d584b036f23f654192a3fbea2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114946/
Test FAILed.

planga82 committed:
[SPARK-29922] Solve compatibility test fail

SparkQA commented:
**[Test build #114960 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114960/testReport)** for PR 26667 at commit [`a11f0ed`](https://github.com/apache/spark/commit/a11f0ed4f7c2beb2b4f9e9e01f5436a615a82271).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19781/
Test PASSed.

SparkQA commented:
**[Test build #114960 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114960/testReport)** for PR 26667 at commit [`a11f0ed`](https://github.com/apache/spark/commit/a11f0ed4f7c2beb2b4f9e9e01f5436a615a82271).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114960/
Test PASSed.

planga82 commented:
Test problems solved and comments implemented
Cc @viirya .

viirya commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -3273,4 +3273,20 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
+  override def visitShowFunctions(ctx: ShowFunctionsContext): LogicalPlan = withOrigin(ctx) {
+    val (userScope, systemScope) = Option(ctx.identifier)
+      .map(_.getText.toLowerCase(Locale.ROOT)) match {
+      case s @ (None | Some("all")) => (true, true)
```
just `case None | Some("all") => (true, true)`? You do not access `s`.

viirya replied:
indentation:

```scala
val (userScope, systemScope) = Option(ctx.identifier)
  .map(_.getText.toLowerCase(Locale.ROOT)) match {
    case None | Some("all") => (true, true)
    ...
  }  
}

viirya reviewed (approved):
two minor comments about style.

planga82 committed:
[SPARK-29922] Implement style comments

SparkQA commented:
**[Test build #114997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114997/testReport)** for PR 26667 at commit [`e630e2c`](https://github.com/apache/spark/commit/e630e2cd7be5ab08852cde6fdb6106d82e89bef4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/19816/
Test PASSed.

SparkQA commented:
**[Test build #114997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/114997/testReport)** for PR 26667 at commit [`e630e2c`](https://github.com/apache/spark/commit/e630e2cd7be5ab08852cde6fdb6106d82e89bef4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/114997/
Test PASSed.

imback82 reviewed (approved):
LGTM

viirya commented:
Thanks! Merging to master.

planga82 commented:
Thanks all!!
